File: markov_decision_process/tictactoe.py
import numpy as np


class TicTacToe:
    def __init__(self):
        self.state_options = ['_', 'O', 'X']

    def state_to_index(self, state):
        state_num = "".join([str(x) for x in state.reshape(-1).tolist()])
        state_num = int(state_num, 3)
        return state_num

    def index_to_state(self, n):
        state = []

        for i in range(9):
            cur = n // 3 ** (8 - i)
            state.append(cur)

            n = n - cur * (3 ** (8 - i))

        return np.reshape(state, (3, 3))

    def print(self, state, with_numbers=False):
        out = ""

        for i, row in enumerate(state):
            for j, col in enumerate(row):
                if col == 0 and with_numbers:
                    out += str((i * 3 + j) + 1) + " "
                else:
                    out += self.state_options[col] + " "
            if i < 2:
                out += "\n"

        print(out)

    def is_done(self, state):

        if np.all(state != 0):
            return True

        for i in range(3):
            if np.all(state[i] == state[i, 0]) and state[i, 0] != 0:
                return True
            if np.all(state[:, i] == state[0, i]) and state[0, i] != 0:
                return True

        if np.all(np.diagonal(state) == state[0, 0]) and state[0, 0] != 0:
            return True

        if np.all(np.diagonal(np.fliplr(state)) == state[0, 2]) and state[0, 2] != 0:
            return True

        return False

    def get_reward_P_trans(self, state_index, u, side=1):
        state = self.index_to_state(state_index)
        reward = 0
        loss_factor = 10
        opp_side = 2 if side == 1 else 1
        P_trans = np.zeros(3 ** 9)

        if self.is_done(state):
            P_trans[state_index] = 1

        elif np.count_nonzero(state == opp_side) > np.count_nonzero(state == side):
            i = u // 3
            j = u % 3

            if state[i, j] != 0:
                P_trans[state_index] = 1
                reward = -1
            else:
                new_state = np.copy(state)
                new_state[i, j] = side
                P_trans[self.state_to_index(new_state)] = 1

        else:
            blanks = np.argwhere(state == 0)
            blanks_count = np.shape(blanks)[0]
            blank_prob = (1 / blanks_count) if blanks_count > 0 else 0

            for blank in blanks:
                blank_state = np.copy(state)
                blank_state[blank[0], blank[1]] = opp_side
                blank_state_index = self.state_to_index(blank_state)
                P_trans[blank_state_index] = blank_prob

        for i in range(3):
            row = state[i, :]
            col = state[:, i]

            if np.all(row == opp_side) or np.all(col == opp_side):
                return - loss_factor, P_trans
            if np.all(row == side) or np.all(col == side):
                return 1, P_trans

        diag = np.diagonal(state)
        if np.all(diag == opp_side):
            return - loss_factor, P_trans
        if np.all(diag == side):
            return 1, P_trans

        diag2 = np.diagonal(np.fliplr(state))
        if np.all(diag2 == opp_side):
            return - loss_factor, P_trans
        if np.all(diag2 == side):
            return 1, P_trans

        return reward, P_trans

    def play(self, state, u, side=1):
        opp_side = 2 if side == 1 else 1
        state = np.copy(state)

        if self.is_done(state):
            return state

        if np.count_nonzero(state == opp_side) > np.count_nonzero(state == side):
            i = u // 3
            j = u % 3

            if state[i, j] == 0:
                state[i, j] = side

        else:
            u = np.random.choice(np.arange(9))
            i = u // 3
            j = u % 3

            while state[i, j] != 0:
                u = np.random.choice(np.arange(9))
                i = u // 3
                j = u % 3

            state[i, j] = opp_side

        return state

    def user_play(self, state, side=1):
        try:
            state = np.copy(state)

            u = input("Position (1 to 9): ")
            u = int(u) - 1

            i = u // 3
            j = u % 3

            while u < 0 or u > 8 or state[i, j] != 0:
                u = input("Position (1 to 9): ")
                u = int(u) - 1

                i = u // 3
                j = u % 3

            state[i, j] = side
            return state

        except ValueError:
            return self.user_play(state, side)

File: cpp/markov_decision_process/ticTacToe.cpp
#include "ticTacToe.h"

using namespace std;
using namespace Eigen;

IOFormat VectorFormat(0, 0, "", "");

TicTacToe::TicTacToe()
{
    this->initialState = Matrix3i::Zero();
}

int TicTacToe::stateToIndex(TTTState state)
{
    stringstream stateStr;
    stateStr << state.format(VectorFormat);
    return stoi(stateStr.str(), nullptr, 3);
}

TTTState TicTacToe::indexToState(int index)
{
    MatrixXi state = MatrixXi::Zero(1, 9);

    for (int i = 0; i < 9; i++)
    {
        int cur = index / (int)pow(3, 8 - i);
        state(0, i) = cur;
        index = index - cur * pow(3, 8 - i);
    }

    return state.reshaped<RowMajor>(3, 3);
}

std::tuple<int, Eigen::VectorXd> TicTacToe::getRewardAndPTrans(int stateIndex, TTTAction u)
{
    TTTState state = indexToState(stateIndex);

    int reward = 0;
    VectorXd PTrans((int)pow(3, 9));

    if (isDone(state))
        PTrans(stateIndex) = 1;

    else if ((state.array() == oppSide).count() > (state.array() == side).count())
    {
        int i = u / 3, j = u % 3;

        if (state(i, j) != 0)
        {
            PTrans(stateIndex) = 1;
            reward = -1;
        }
        else
        {
            TTTState newState = state;
            newState(i, j) = side;
            PTrans(stateToIndex(newState)) = 1;
        }
    }
    else
    {
        Matrix3i blanks = (state.array() == 0).cast<int>();
        int blanksCount = blanks.count();
        float blanksProb = blanksCount > 0 ? (1. / blanksCount) : 0;

        for (int bI = 0; bI < 3; bI++)
        {
            for (int bJ = 0; bJ < 3; bJ++)
            {
                if (blanks(bI, bJ))
                {
                    TTTState blankState = state;
                    blankState(bI, bJ) = oppSide;
                    int blankStateIndex = stateToIndex(blankState);
                    PTrans(blankStateIndex) = blanksProb;
                }
            }
        }
    }

    for (int i = 0; i < 3; i++)
    {
        Vector3i row = state.row(i);
        Vector3i col = state.col(i);

        if ((row.array() == oppSide).all() || (col.array() == oppSide).all())
            return {-10, PTrans};

        if ((row.array() == side).all() || (col.array() == side).all())
            return {1, PTrans};
    }

    Vector3i diag = state.diagonal();
    if ((diag.array() == oppSide).all())
        return {-10, PTrans};
    if ((diag.array() == side).all())
        return {1, PTrans};

    Vector3i diag2 = state.rowwise().reverse().diagonal();
    if ((diag2.array() == oppSide).all())
        return {-10, PTrans};
    if ((diag2.array() == side).all())
        return {1, PTrans};

    return {reward, PTrans};
}

bool TicTacToe::isDone(TTTState state)
{
    if ((state.array() != 0).all())
        return true;

    for (int i = 0; i < 3; i++)
    {
        if (state(i, 0) != 0 && (state.row(i).array() == state(i, 0)).all())
            return true;

        if (state(0, i) != 0 && (state.col(i).array() == state(0, i)).all())
            return true;
    }

    if (state(0, 0) != 0 && (state.diagonal().array() == state(0, 0)).all())
        return true;

    if (state(0, 2) != 0 && (state.rowwise().reverse().diagonal().array() == state(0, 2)).all())
        return true;

    return false;
}

TTTState TicTacToe::play(TTTState oldState, TTTAction u)
{
    int side = 1, oppSide = 2;
    TTTState state = oldState;

    if (isDone(state))
        return oldState;

    if ((state.array() == oppSide).count() > (state.array() == side).count())
    {
        int i = u / 3, j = u % 3;

        if (state(i, j) == 0)
            state(i, j) = side;
    }

    else
    {
        int i = 0, j = 0;

        do
        {
            TTTAction u = TTTAction(rand() % 9);
            i = u / 3, j = u % 3;
        } while (state(i, j) != 0);

        state(i, j) = oppSide;
    }

    return state;
}

string TicTacToe::displayGame(TTTState state)
{
    string out = "";

    for (int i = 0; i < 3; i++)
    {
        for (int j = 0; j < 3; j++)
        {
            string character = "";
            switch (state(i, j))
            {
            case TTTStateOptions::X:
                character = "X";
                break;

            case TTTStateOptions::O:
                character = "O";
                break;

            default:
                character = "_";
                break;
            }

            out.append(character + " ");
        }
        if (i < 2)
            out.append("\n");
    }

    return out;
}

TTTState TicTacToe::userPlay(TTTState oldState)
{
    TTTState state = oldState;
    string uStr;
    int u, i, j;

    do
    {
        cout << "Position (1 to 9): ";
        getline(cin, uStr);
        if (uStr.length() == 0 || !isdigit(uStr[0]))
        {
            u = -1;
            continue;
        }
        u = uStr[0] - '0' - 1;
        i = u / 3, j = u % 3;

    } while (u < 0 || u > 8 || state(i, j) != 0);

    state(i, j) = oppSide;
    return state;
}

void TicTacToe::playMatch(int numEps, MatrixXd policy)
{
    for (int i = 0; i < numEps; i++)
    {
        TTTState state = Matrix3i::Zero();

        while (!isDone(state))
        {
            state = userPlay(state);
            cout << displayGame(state) << endl;
            cout << "-------" << endl;

            if (isDone(state))
                break;

            int stateIndex = stateToIndex(state);
            TTTAction u;
            policy.row(stateIndex).maxCoeff(&u);

            state = play(state, u);
            cout << displayGame(state) << endl;
            cout << "-------" << endl;
        }

        cout << "DONE" << endl;
    }
}